csrc/dynamic_transform.cpp

```diff
@@ -0,0 +1,510 @@
+  auto all_stmts = StmtSort::getStmts(info_.fusion(), false);
+  for (auto stmt : all_stmts) {
+    if (stmt->isA<Val>()) {
+      std::cerr << "mutate: " << stmt->toString() << std::endl;
+      mutate(stmt);
+    }
+  }
+}
+
+void DynamicTransformConcretizer::concretizeReshape() {
+  DEBUG_PRINT_SCOPE();
+
+  // Concretize each reshape op.
+  for (const auto& kv : info_.getReshapeTransforms()) {
+    auto incomplete_out_tv = kv.first;
+    const auto view_analysis = kv.second;
+
+    std::cerr << "view: " << view_analysis.toString() << std::endl;
+
+    auto inp_tv = ir_utils::producerTvsOf(incomplete_out_tv).at(0);
+
+    auto concrete_reshape_out_tv = reshape(inp_tv, view_analysis);
+
+    std::cerr << "concrete view out: " << concrete_reshape_out_tv->toString()
+              << ", expr: " << concrete_reshape_out_tv->definition()->toString()
+              << std::endl;
+
+    if (inp_tv != concrete_reshape_out_tv->definition()->input(0)) {
+      std::cerr << "Addnl expr: "
+                << concrete_reshape_out_tv->definition()
+                       ->input(0)
+                       ->definition()
+                       ->toString();
+    }
+
+    // Replace the old tensor with the new concretized tensor
+    for (auto use_of_old_tv : incomplete_out_tv->uses()) {
+      std::cerr << "Before replacement: " << use_of_old_tv->toString();
+      auto new_use = ir_utils::replaceValInExpr(
+          use_of_old_tv, incomplete_out_tv, concrete_reshape_out_tv);
+      std::cerr << "After replacement: " << new_use->toString();
+    }
+
+    if (incomplete_out_tv->isFusionOutput()) {
+      incomplete_out_tv->fusion()->removeOutput(incomplete_out_tv);
+      incomplete_out_tv->fusion()->addOutput(concrete_reshape_out_tv);
+      incomplete_out_tv->fusion()->removeVal(incomplete_out_tv);
+    }
+  }
+}
+
+namespace {
+
+bool hasSymbolicAxis(const std::vector<IterDomain*>& ids) {
+  return std::any_of(ids.begin(), ids.end(), [](IterDomain* id) {
+    return id->getIterType() == IterType::Symbolic;
+  });
+}
+
+} // namespace
+
+void DynamicTransformConcretizer::mutate(TensorView* tv) {
+  auto propagated = propagateFromProducerToConsumer(tv);
+
+  if (!propagated) {
+    return;
+  }
+
+  // At this point, there should be no expr beyond rfactor root
+  TORCH_INTERNAL_ASSERT(
+      tv->domain()->domain() == tv->getMaybeRFactorDomain(),
+      "Invalid tensor: ",
+      tv->toString());
+
+  // If it has a rfactor root domain, instead of using the
+  // OptOutMutator interface, replay the exprs with the mutated root
+  // IDs and then register the mutated rf domain IDs. This simplifies
+  // the logic as new IterDomains are automatically generated by the
+  // replay.
+  if (tv->hasRFactor()) {
+    std::unordered_map<IterDomain*, IterDomain*> replacement_map;
+    for (auto root_id : tv->getRootDomain()) {
+      auto mutated = maybeMutated(root_id)->as<IterDomain>();
+      TORCH_INTERNAL_ASSERT(mutated != root_id);
```

Comment:
This seems specific to reshape and might not be the same for operations where not all iter domains are impacted by the semantics of the alias like operation.

Reply:
Updated this part. Instead of using `ReplayTransformations`, just changed to traverse all ID exprs from root to rfactor and update IterType when necessary.